packages/appium/lib/config-file.js

```diff
@@ -0,0 +1,389 @@
+  log.debug(`Attempting to parse ${filepath} as YAML`);
+  return yaml.parse(content);
+}
+
+/**
+ * A cache of the raw config file (a JSON string) at a filepath.
+ * This is used for better error reporting.
+ * Note that config files needn't be JSON, but it helps if they are.
+ * @type {Map<string,RawJson>}
+ */
+const rawConfig = new Map();
+
+/**
+ * Custom JSON loader that caches the raw config file (for use with `better-ajv-errors`).
+ * If it weren't for this cache, this would be unnecessary.
+ * @type {import('lilconfig').LoaderSync}
+ */
+function jsonLoader (filepath, content) {
+  log.debug(`Attempting to parse ${filepath} as JSON`);
+  rawConfig.set(filepath, content);
+  return JSON.parse(content);
+}
+
+/**
+ * Loads a config file from an explicit path
+ * @param {LilconfigAsyncSearcher} lc - lilconfig instance
+ * @param {string} filepath - Path to config file
+ */
+async function loadConfigFile (lc, filepath) {
+  log.debug(`Attempting to load config at filepath ${filepath}`);
+  try {
+    // removing "await" will cause any rejection to _not_ be caught in this block!
+    return await lc.load(filepath);
+  } catch (err) {
+    if (err.code === 'ENOENT') {
+      err.message = `Config file not found at user-provided path: ${filepath}`;
+    } else if (err instanceof SyntaxError) {
+      // generally invalid JSON
+      err.message = `Config file at user-provided path ${filepath} is invalid:\n${err.message}`;
+    }
+    throw err;
+  }
+}
+
+/**
+ * Searches for a config file
+ * @param {LilconfigAsyncSearcher} lc - lilconfig instance
+ */
+async function searchConfigFile (lc) {
+  log.debug('No config file specified; searching...');
+  const result = await lc.search();
+  if (!result?.filepath) {
+    log.debug('Could not find a config file');
+  }
+  return result;
+}
+
+/**
+ * Reads an Appium config file; searches for one if no `filepath` specified.
+ * @param {string} [filepath] - Explicit path to config file
+ * @returns {Promise<ReadConfigFileResult>} Empty object if not found.
+ */
+async function findConfigFile (filepath) {
+  const lc = lilconfig('appium', {
+    loaders: {
+      '.yaml': yamlLoader,
+      '.yml': yamlLoader,
+      '.json': jsonLoader,
+      noExt: jsonLoader,
+    },
+  });
+  return (
+    (await (filepath ? loadConfigFile(lc, filepath) : searchConfigFile(lc))) ??
+    {}
+  );
+}
+
+/**
+ * Given an object, validates it against the Appium config schema.
+ * If errors occur, the returned array will be non-empty.
+ * @todo Provide API to validate against driver/plugin schemas.
+ * @public
+ * @param {any} value - The value (hopefully an object) to validate against the schema
+ * @returns {import('ajv').ErrorObject[]} Array of errors, if any.
+ */
+export function validateConfig (value) {
+  const validator = getValidator(APPIUM_CONFIG_SCHEMA_ID);
+  if (!validator) {
+    throw new Error(
+      `Could not find schema with ID: ${APPIUM_CONFIG_SCHEMA_ID}`,
+    );
+  }
+  return !validator(value) &&
```

Comment:
NIT: can we put `!validator(value) && _.isArray(validator.errors)` on the same line so the semantics of this ternary operator are more transparent?